src/pip/_internal/commands/install.py

```diff
@@ -546,63 +536,83 @@ def site_packages_writable(**kwargs):
-    If use_user_site is True, it is checked for compatibility with other
-    options.
-    If use_user_site is None, the default behaviour depends on the environment,
-    which is provided by the other arguments.
-    """
-    # In some cases (config from tox), use_user_site can be set to an integer
-    # rather than a bool, which 'use_user_site is False' wouldn't catch.
-    if (use_user_site is not None) and (not use_user_site):
-        logger.debug("Non-user install by explicit request")
-        return False
+    If use_user_site is None, the behavior is decided upon other
+    arguments and the environment.
 
-    if use_user_site:
-        if prefix_path:
-            raise CommandError(
-                "Can not combine '--user' and '--prefix' as they imply "
-                "different installation locations"
-            )
-        if virtualenv_no_global():
+    Compatibility of installation location options and the environment
+    is also asserted in this function.
+    """
+    # Check for incompatible options
+    locations = {"--target": target_dir, "--user": use_user_site,
+                 "--root": root_path, "--prefix": prefix_path}
+    destinations = [k for k, v in locations.items() if v]
+    if len(destinations) > 1:
+        last, rest = destinations[-1], ", ".join(destinations[:-1])
+        raise CommandError("Different installation locations are implied by "
+                           "{} and {}.".format(rest, last))
+
+    if target_dir:
+        if os.path.exists(target_dir) and not os.path.isdir(target_dir):
+            raise InstallationError("Target path exists but is not "
+                                    "a directory: {}".format(target_dir))
+        if not test_writable_dir(target_dir):
             raise InstallationError(
-                "Can not perform a '--user' install. User site-packages "
-                "are not visible in this virtualenv."
-            )
-        logger.debug("User install by explicit request")
-        return True
-
-    # If we are here, user installs have not been explicitly requested/avoided
-    assert use_user_site is None
-
-    # user install incompatible with --prefix/--target
-    if prefix_path or target_dir:
-        logger.debug("Non-user install due to --prefix or --target option")
+                "Cannot write to target directory: {}".format(target_dir))
+        logger.debug("Non-user install due to specified target directory")
         return False
-
-    # If user installs are not enabled, choose a non-user install
-    if not site.ENABLE_USER_SITE:
-        logger.debug("Non-user install because user site-packages disabled")
+    if prefix_path:
+        if not test_writable_dir(prefix_path):
+            raise InstallationError(
+                "Cannot write to prefix path: {}".format(prefix_path))
+        logger.debug("Non-user install due to specified prefix path")
         return False
 
```

Comment:
For the purposes of readability, I like labeling conditionals with var names:
```suggestion
    is_user_site_disabled_for_security = site.ENABLE_USER_SITE is None

```
It then makes more sense in the code.